workflow/source.go

```diff
@@ -129,59 +141,98 @@ func (s Source) Validate() error {
 }
 
 func (s Source) GetOutputLocation() (string, error) {
-    if s.Output != nil {
-        if len(s.Inputs) > 1 && !isYAMLFile(*s.Output) {
-            return "", fmt.Errorf("when merging multiple inputs, output must be a yaml file")
-        }
-        return *s.Output, nil
-    }
-
-    if len(s.Inputs) == 1 && len(s.Overlays) == 0 {
-        return s.handleSingleInput()
-    }
-
-    return s.generateOutputPath()
-}
-
-func (s Source) handleSingleInput() (string, error) {
-    input := s.Inputs[0].Location
-    switch getFileStatus(input) {
-    case fileStatusLocal:
-        return input, nil
-    case fileStatusNotExists:
-        return "", fmt.Errorf("input file %s does not exist", input)
-    case fileStatusRemote, fileStatusRegistry:
-        return s.generateRegistryPath(input)
-    default:
-        return "", fmt.Errorf("unknown file status for %s", input)
-    }
+	if s.Output != nil {
+		if len(s.Inputs) > 1 && !isYAMLFile(*s.Output) {
+			return "", fmt.Errorf("when merging multiple inputs, output must be a yaml file")
+		}
+		return *s.Output, nil
+	}
+
+	return s.generateOutputPath()
 }
 
 func (s Source) generateRegistryPath(input string) (string, error) {
-    ext := filepath.Ext(input)
-    if ext == "" {
-        ext = ".yaml"
-    }
-    hash := fmt.Sprintf("%x", sha256.Sum256([]byte(input)))
-    return filepath.Join(GetTempDir(), fmt.Sprintf("registry_%s%s", hash[:6], ext)), nil
+	ext := filepath.Ext(input)
+	hash := fmt.Sprintf("%x", sha256.Sum256([]byte(input)))
+
+	if ext == "" {
+		resolvedExtension := s.getRemoteResolvedExtension(input)
+
+		if resolvedExtension != InputFileTypeUnknown {
+			return filepath.Join(GetTempDir(), fmt.Sprintf("registry_%s.%s", hash[:6], resolvedExtension)), nil
+		}
+	}
+
+	// Check if the extension is supported
+	if ext != ".yaml" && ext != ".yml" && ext != ".json" {
+		ext = ".yaml"
+	}
+	return filepath.Join(GetTempDir(), fmt.Sprintf("registry_%s%s", hash[:6], ext)), nil
+}
+
+// Attempts to fetch the remote file and determine the format (yaml / json)
+// based on the contents of the file
+func (s Source) getRemoteResolvedExtension(input string) InputFileType {
+	res, err := http.Get(input)
```

Comment:
I don't think this works for a lot of cases:

1. Some remote files are blocked by authentication headers - these are provided alongside the input location.
2. This `getRemoteResolvedExtension(input)` is also used with registry URIs.

I think we need to delegate to `ResolveDocument` which currently lives within the speakeasy repo.

Reply:
Okay, sounds good. I will do some investigation into this 👍🏻 

Aside from this fetching logic to be stripped out and replaced with `ResolveDocument`, the approach is generally sound?

Reply:
Yep; definitely!